gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/experimental/internal/configureNativeCompilerCaching.kt

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright 2020-2023 JetBrains s.r.o. and respective authors and developers.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE.txt file.
+ */
+
+package org.jetbrains.compose.experimental.internal
+
+import org.gradle.api.Project
+import org.jetbrains.compose.internal.KOTLIN_MPP_PLUGIN_ID
+import org.jetbrains.compose.internal.mppExt
+import org.jetbrains.compose.internal.utils.KGPPropertyProvider
+import org.jetbrains.compose.internal.utils.configureEachWithType
+import org.jetbrains.kotlin.gradle.plugin.getKotlinPluginVersion
+import org.jetbrains.kotlin.gradle.plugin.mpp.KotlinNativeTarget
+import org.jetbrains.kotlin.konan.target.presetName
+
+private const val PROJECT_CACHE_KIND_PROPERTY_NAME = "kotlin.native.cacheKind"
+private const val COMPOSE_NATIVE_MANAGE_CACHE_KIND = "compose.kotlin.native.manageCacheKind"
+private const val NONE_VALUE = "none"
+
+internal fun Project.configureNativeCompilerCaching() {
+    if (findProperty(COMPOSE_NATIVE_MANAGE_CACHE_KIND) == "false") return
+
+    plugins.withId(KOTLIN_MPP_PLUGIN_ID) {
+        val kotlinVersion = kotlinVersionNumbers(this)
+        mppExt.targets.configureEachWithType<KotlinNativeTarget> {
+            assertCacheKindUserValueIsNotNone()
+            configureTargetCompilerCache(kotlinVersion)
+        }
+    }
+}
+
+private fun KotlinNativeTarget.assertCacheKindUserValueIsNotNone() {
+    // To determine cache kind KGP checks kotlin.native.cacheKind.<PRESET_NAME> first, then kotlin.native.cacheKind
+    // For each property it tries to read Project.property, then checks local.properties
+    // See https://github.com/JetBrains/kotlin/blob/d4d30dcfcf1afb083f09279c6f1ba05031efeabb/libraries/tools/kotlin-gradle-plugin/src/common/kotlin/org/jetbrains/kotlin/gradle/plugin/PropertiesProvider.kt#L416
+    val cacheKindProperties = listOf(targetCacheKindPropertyName, PROJECT_CACHE_KIND_PROPERTY_NAME)
+    val propertyProviders = listOf(
+        KGPPropertyProvider.GradleProperties(project),
+        KGPPropertyProvider.LocalProperties(project)
+    )
```

Comment:
When `kotlin.native.cacheKind` accidentally exists in local.properties - it is mistake in project configuraion.
I don't know how exactly better to behave.
I think we can just skip this case. But, I don't have a strong opinion.

Reply:
Oops, sorry. Didn't notice about KGP comment. Well - it is the best way to do so!
Thanks!

Reply:
Oops, sorry. Didn't notice about KGP comment. Well - it is the best way to do so!
Thanks!